### DISCUSSION TITLE
python-poppler now uses meson-python

### DISCUSSION THREAD

**Discussion by cbrunet** [mesonbuild/meson-python#356] [Show and tell]
Just to let you know I switched from cmake to meson to build [python-poppler](https://github.com/cbrunet/python-poppler).

This simple project can be a good example how to build a python module using pybind11 and meson.
It also show integration with tox to run tests.

**Comment by rgommers** [comment]
Thanks for sharing @cbrunet! That looks pretty clean indeed.

**Comment by dnicolodi** [comment]
Thank you for sharing. I see a few anti-patterns in the Meson build definition. I'll send a PR when I find a minute.

**Reply by cbrunet** [reply]
I will probably do a release soon, due to incompatibilities induced by the new meson-python version. Your suggestions would be more that welcome!

**Reply by dnicolodi** [reply]
I would like to know more about the backward compatibility break you observe with meson-python 0.13.0. The release was intended to be completely backward compatible with the previous release, as far as I can remember. Looking at the python-poppler git commits I see a fix related to the Python package being split between purelib and platlib. This is only meson-python informing you that the wheel package that would be created is broken. The bug was in python-poppler, but it was not observed. meson-python just tells you this.

**Reply by cbrunet** [reply]
This was reported as an error, and I was told the pip installation was broken because of that (https://github.com/cbrunet/python-poppler/issues/74). However, I do not distribute wheel (yet), so I'm confused. And tests were passing when I released it, therefore the installation was working.

**Reply by dnicolodi** [reply]
What confuses you? pip can install from source.

The fact that the produced wheels apparently worked without issue is exactly what I mean with "the bug was not observed". Placing components for the same import package in both platlib and purelib is undefined in the wheel specification and almost certainly broken on platforms where these two directories are distinct. However, on the most common platforms these are the same directory, thus no one notices. This is the reason the check has been added to meson-python: to make package authors notice the problem.

**Reply by rgommers** [reply]
> However, I do not distribute wheel (yet)

It doesn't matter if you're the one distributing a wheel or not; if anyone types `pip install python-poppler`, `pip` will grab the sdist and try to build a wheel from it itself.

>  therefore the installation was working.

It was working on most systems, but silently broken on some of the more niche Linux distros which split `platlib` and `purelib` into separate directories. So what was happening is that part of your package is installed to one root directory, part to another root directory, and whether it worked or not depended on whether platlib/purelib were pointing to the same directory. (and yes, this stuff is conceptually very weird).

With the meson-python 0.13.0 release, it went from "silently broken on a few distros" to "loudly broken everywhere". 

In hindsight, we could have emitted a warning for one release. But on the other hand, everyone tends to ignore those so flagging the problem loudly sooner rather than later is what we went for.

**Comment by AshleaPinntT** [comment]
Wow, it's great to see how open-source projects like python-poppler continue to evolve and improve over time!

**Comment by KellyLehning9h** [comment]
That's great to hear! Switching to meson can certainly make the build process more efficient. And thanks for sharing your project as an example, it's always helpful to have more resources available for building Python modules using pybind11 and meson. And using tox for running tests is also a good practice. Well done!
